File: example/ThumbnailGenerator/ThumbnailGenerator/ViewController.m
//
//  ViewController.m
//  ThumbnailGenerator
//
//  Created by Daniel on 2023/02/15.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self extractImagesFromMovie];
    NSLog (@"Done.");
}

    
-(void)extractImagesFromMovie {
    NSURL* url = [NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:@"test" ofType:@"mp4"]];

    NSMutableDictionary* dict = [NSMutableDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithBool:YES] ,
                                     AVURLAssetPreferPreciseDurationAndTimingKey ,
                                    [NSNumber numberWithInt:0],
                                    AVURLAssetReferenceRestrictionsKey, nil];

    AVURLAsset* avUrlAsset = [[AVURLAsset alloc] initWithURL:url options:dict];

    // set the generator
    AVAssetImageGenerator* generator = [AVAssetImageGenerator assetImageGeneratorWithAsset:avUrlAsset];
    generator.requestedTimeToleranceBefore = kCMTimeZero;
    generator.requestedTimeToleranceAfter = kCMTimeZero;
    generator.maximumSize = CGSizeMake(192, 144);

    // look for the video track
    AVAssetTrack* videoTrack = nil;
    bool foundTrack = NO;

    for (AVAssetTrack* track in avUrlAsset.tracks) {
        if ([track.mediaType isEqualToString:@"vide"]) {
            if (foundTrack) {
                NSLog (@"Error - - - more than one video tracks");
                return;
            } else {
                videoTrack = track;
                foundTrack = YES;
            }
        }
    }
    
    if (videoTrack == nil) {
        NSLog (@"Error - - No Video Tracks at all");
        return;
    }

    // set the number of frames in the movie
    int frameRate = videoTrack.nominalFrameRate;
    float value = avUrlAsset.duration.value;
    float timeScale = avUrlAsset.duration.timescale;
    float totalSeconds = value / timeScale;
    int totalFrames = totalSeconds * frameRate;
    int timeValuePerFrame = avUrlAsset.duration.timescale / frameRate;
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSMutableArray *times=[[NSMutableArray alloc] init];

    // get each frame
    for (int k=1; k<=totalFrames; k++) {
        unsigned long timeValue = timeValuePerFrame * k;
        unsigned int timeStampMsec = timeValue*1000 / timeScale;

        // Set times every seconds : 1st frame + frame at 1000*n msec.
        if (k == 1 || timeStampMsec % 1000 == 0) {
            CMTime frameTime;
            frameTime.value = timeValue;
            frameTime.timescale = timeScale;
            frameTime.flags = avUrlAsset.duration.flags;
            frameTime.epoch = avUrlAsset.duration.epoch;
            NSLog (@"Set time at Frame %d Time Stamp %lld", k, frameTime.value*1000/frameTime.timescale);
            [times addObject:[NSValue valueWithCMTime:frameTime]];
        }
    }
    
    [generator generateCGImagesAsynchronouslyForTimes:times
                                    completionHandler:^(CMTime requestedTime,
                                                        CGImageRef image,
                                                        CMTime actualTime,
                                                        AVAssetImageGeneratorResult result,
                                                        NSError *error) {
        NSString *requestedTimeString = (__bridge NSString *) CMTimeCopyDescription (NULL, requestedTime);
        NSString *actualTimeString = (__bridge NSString *) CMTimeCopyDescription(NULL, actualTime);

        if (result == AVAssetImageGeneratorSucceeded) {
            if (actualTime.value != requestedTime.value) {
                UIImage *img = [UIImage imageWithCGImage:image];
                NSString *filePath = [[paths objectAtIndex:0] stringByAppendingPathComponent:[NSString stringWithFormat:@"%lld_sec.png",requestedTime.value/requestedTime.timescale]];
    
                NSData *data = [NSData dataWithData:UIImagePNGRepresentation(img)];
                NSError *writeError = nil;
    
                [data writeToFile:filePath options:NSDataWritingAtomic error:&writeError];
    
                if (writeError) {
                    NSLog(@"Error writing file: %@", writeError);
                } else {
                    NSLog(@"Generated '%lld_sec.png'", requestedTime.value/requestedTime.timescale);
                }
            }
        } else if (result == AVAssetImageGeneratorFailed) {
            NSLog(@"Failed with error: %@", [error localizedDescription]);
        } else if (result == AVAssetImageGeneratorCancelled) {
            NSLog(@"Canceled");
        }
    }];
}


@end

File: example/ThumbnailGeneratorSwift/ThumbnailGeneratorSwift/ViewController.swift
//
//  ViewController.swift
//  ThumbnailGeneratorSwift
//
//  Created by Daniel on 2023/02/16.
//

import UIKit
import AVKit

class ViewController: UIViewController {
    
    var imageView = UIImageView(frame: CGRectMake(0,0,0,0))

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        imageView.backgroundColor = UIColor.black
        self.view.addSubview(imageView)
        test()
    }

    func test() {
        Task {
            await extractImagesFromavAsset()
        }
    }
    
    func extractImagesFromavAsset() async {
        let videoURL = URL(fileURLWithPath: Bundle.main.path(forResource: "test", ofType: "mp4") ?? "")
        let avAsset = AVAsset(url: videoURL)

        // set the generator
        let generator = AVAssetImageGenerator(asset: avAsset)
        generator.requestedTimeToleranceBefore = .zero
        generator.requestedTimeToleranceAfter = .zero
        generator.maximumSize = CGSize(width: 192, height: 144)
        
        // look for the video track
        let tracks = try? await avAsset.load(.tracks)
        var foundTrack : AVAssetTrack? = nil;
        for i in 0..<Int(tracks!.count) {
            if tracks![i].mediaType.rawValue == "vide" {
                foundTrack = tracks![i]
                break
            }
        }

        if foundTrack == nil {
            print("Error - - No Video Tracks")
            return
        }

        // set the information about frames in the avAsset
        let frameRate = try? await Int(foundTrack?.load(.nominalFrameRate) ?? 0)

        let value = try? await Float(avAsset.load(.duration).value)
        let timeScale = try? await Float(avAsset.load(.duration).timescale)
        let flags = try? await avAsset.load(.duration).flags
        let epoch = try? await avAsset.load(.duration).epoch

        let totalSeconds = value! / timeScale!
        let totalFrames = Int(totalSeconds * Float(frameRate!))
        let timeValuePerFrame = Int(timeScale!) / frameRate!
        
        print("total frames \(totalFrames)")
        
        // get each frame
        var times: [AnyHashable] = []
        for k in 1...totalFrames {
            let timeValue = UInt(timeValuePerFrame * k)
            let timeStampMsec = UInt(timeValue * 1000) / UInt(timeScale!)

            // Set times every seconds : 1st frame + frame at 1000*n msec.
            if k == 1 || timeStampMsec % 1000 == 0 {
                var frameTime: CMTime! = CMTime()
                frameTime.value = CMTimeValue(timeValue)
                frameTime.timescale = CMTimeScale(timeScale!)
                frameTime.flags = flags!
                frameTime.epoch = epoch!
                print(String(format: "Set time at Frame %d Time Stamp %lld", k, CMTimeScale(frameTime.value * 1000) / frameTime.timescale))
                times.append(NSValue(time: frameTime))
            }
        }

        // Show & store thumbnails every seconds.
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).map(\.path)
        if let times = times as? [NSValue] {
            generator.generateCGImagesAsynchronously(forTimes: times) {
                requestedTime, image, actualTime, result, error in
                    if result == .succeeded && actualTime.value != requestedTime.value {
                        var img: UIImage? = nil
                        if let image {
                            img = UIImage(cgImage: image)
                            if let data = img?.pngData() {
                                // Show thumbnail
                                DispatchQueue.main.async {
                                    self.imageView.frame = CGRectMake((self.view.frame.size.width / 2) - (img!.size.width / 2), (self.view.frame.size.height / 2) - (img!.size.height / 2), img!.size.width, img!.size.height);
                                    self.imageView.image = img
                                }
                                // Store thumbnail
                                let fileName = String(format: "%lld_sec.png", requestedTime.value / Int64(requestedTime.timescale))
                                let filePath = URL(fileURLWithPath: paths[0]).appendingPathComponent(fileName)
                                try? data.write(to: filePath)
                                print("Store file : \(fileName)")
                            }
                        }
                    }
            }
        }
    }

}

